cmd/app/metricsmonitor.go

```diff
@@ -0,0 +1,111 @@
+package app
+
+import (
+	"bytes"
+	"fmt"
+	"math/rand"
+	"net/http"
+	"reflect"
+	"runtime"
+	"strconv"
+	"time"
+)
+
+const reportInterval = 10
+const pollInterval = 2
+
+const typeGauge = "gauge"
+const typeCounter = "counter"
+
+type metrics struct {
+	PollCount   int
+	RandomValue int
+	runtime.MemStats
+}
+
+func sendMetric(client http.Client, metricType string, metricName string, metricValue string) {
+	var body = []byte(`{"message":"Hello"}`)
+	request, err := http.NewRequest(http.MethodPost, "http://localhost:8080/update/"+metricType+"/"+metricName+"/"+metricValue, bytes.NewBuffer(body))
+	if err != nil {
+		fmt.Println(err)
+	}
+	request.Header.Set("Content-Type", "application/json; charset=UTF-8")
+
+	resp, errC := client.Do(request)
+	if errC != nil {
+		fmt.Println(errC)
+	}
+	defer resp.Body.Close()
```

Comment:
При выключенном сервере агент падает по панике в этом месте:
```
Post "http://localhost:8080/update/Alloc/gauge/161928": dial tcp [::1]:8080: connectex: No connection could be made because the target machine actively refused it.
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x0 pc=0x770a65]

goroutine 1 [running]:
yalerting/cmd/app.sendMetric({{0x0, 0x0}, 0x0, {0x0, 0x0}, 0x0}, {0x7dcdb5, 0x5}, {0x7dce9b, 0x5}, ...)
        D:/work/src/github.com/rmvl/metric_alerting/cmd/app/metricsmonitor.go:38 +0x565
yalerting/cmd/app.MonitorMetrics()
        D:/work/src/github.com/rmvl/metric_alerting/cmd/app/metricsmonitor.go:99 +0xbb7
main.main()
        D:/work/src/github.com/rmvl/metric_alerting/cmd/agent/main.go:6 +0x25
```

Потому что после получения ошибки от Do необходимо выходить из функции, при этом полученную ошибку следует пробросить вверх, так как ее наличие может повлиять на дальнейшие вычисления.

Reply:
defer в конце функции не имеет смысла, так как основное его назначение - гарантировать, что функция этого оператора выполнится после всех вычислений текущей функции. Ну а в конце функции мы и так находимся после всех вычислений.